### PR TITLE
Add new 2D (hemisphere) normal map generation tool and a working example.

### PR DESCRIPTION
**Pull request by jnmartin84**
I implemented a 3D vector normal map to 2D hemisphere normal map conversion tool as a modification of the original dcbumpgen tool, currently named "dchemigen."

I also have a working example with royalty-free texture and normal map created by "Lobotomy" (from the Discord).

I am using this as part of my work exploring bump-mapping in the Doom 64 engine and would like to share it so everyone can benefit.

### CODE REVIEW

File: utils/dchemigen/readpng.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
+/* typedef unsigned int uint32;
```
Comment (gyrovorbis) [inline utils/dchemigen/readpng.h:1]: Can you make sure this file has a KOS copyright header similar to readpng.c and the rest?

File: utils/dchemigen/get_image_jpg.c [hunk 2]
```diff
@@ -0,0 +1,124 @@
+
```
Comment (gyrovorbis) [inline utils/dchemigen/get_image_jpg.c:1]: Needs a KOS copyright header.

File: utils/dchemigen/get_image.h [hunk 3]
```diff
@@ -0,0 +1,17 @@
+
```
Comment (gyrovorbis) [inline utils/dchemigen/get_image.h:1]: Needs KOS copyright header.

File: utils/dchemigen/get_image.c [hunk 4]
```diff
@@ -0,0 +1,18 @@
+#include <stdio.h>
```
Comment (gyrovorbis) [inline utils/dchemigen/get_image.c:1]: Needs KOS copyright header.

File: utils/dchemigen/dchemigen.c [hunk 5]
```diff
@@ -0,0 +1,185 @@
+#include <stdio.h>
```
Comment (gyrovorbis) [inline utils/dchemigen/dchemigen.c:1]: Needs KOS copyright header.

File: examples/dreamcast/pvr/normalmap/main.cc [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+      documentation would be appreciated but is not required.
+
+   2. Altered source versions must be plainly marked as such, and 
+      must not be misrepresented as being the original software.
+
+   3. This notice may not be removed or altered from any source 
+      distribution.
+ */
+
+/*
```
Comment (gyrovorbis) [inline examples/dreamcast/pvr/normalmap/main.cc:25]: You sure you don't want to put your name or anything on this file as an AUTHOR? Maybe even on this comment block to preserve the original one? Cuz you did do a lot that wasn't here...

**Review by gyrovorbis — COMMENTED** [review]
Looks great, just a couple of style things I have to ensure for compliance!

File: examples/dreamcast/pvr/normalmap/main.cc (hunk 6 above)
Comment (jnmartin84) [reply examples/dreamcast/pvr/normalmap/main.cc:25]: The only work I did on this file was add the comment below that block about Lobotomy modifying the original example it is based on.

File: utils/dchemigen/dchemigen.c (hunk 5 above)
Comment (jnmartin84) [reply utils/dchemigen/dchemigen.c:1]: This one I can say I worked on and add a header to, definitely.

File: utils/dchemigen/get_image.c (hunk 4 above)
Comment (jnmartin84) [reply utils/dchemigen/get_image.c:1]: This is missing header where I lifted it from (dcbumpgen).

File: utils/dchemigen/get_image.h (hunk 3 above)
Comment (jnmartin84) [reply utils/dchemigen/get_image.h:1]: This is missing header where I lifted it from (dcbumpgen).

File: utils/dchemigen/get_image_jpg.c (hunk 2 above)
Comment (jnmartin84) [reply utils/dchemigen/get_image_jpg.c:1]: This is missing header where I lifted it from (dcbumpgen).

File: utils/dchemigen/readpng.h (hunk 1 above)
Comment (jnmartin84) [reply utils/dchemigen/readpng.h:1]: This is missing header where I lifted it from (dcbumpgen).

**Comment by gyrovorbis** [comment]
@QuzarDC Did you wind up figuring out how we should handle attribution here for file that got copy + pasted from files that got copy + pasted from we-have-no-idea-where? 

This PR should be good to go, other than that! (I will test on DreamSDK, though)

**Comment by TapamN** [comment]
pvrtex already supports converting cartesian normal maps to the PVR's spherical coordinates (using --format normal), and can also generate mipmaps and VQ compress the result. If there's a bug in pvrtex, or it lacks a feature that dchemigen has, I think it makes more sense to add dchemigen's features/improvements to pvrtex than add another texture utility to KOS.

**Comment by jnmartin84** [comment]
@gyrovorbis did we know this? I didn't know this so I guess just forget about mine and I'll just use it for my stuff with the bc5 compression

**Comment by TapamN** [comment]
I didn't mention it when I posted pvrtex on dcemu, but it's listed as an option in the readme.

**Comment by QuzarDC** [comment]
> pvrtex already supports converting cartesian normal maps to the PVR's spherical coordinates (using --format normal), and can also generate mipmaps and VQ compress the result. If there's a bug in pvrtex, or it lacks a feature that dchemigen has, I think it makes more sense to add dchemigen's features/improvements to pvrtex than add another texture utility to KOS.

I'd agree generally with the approach of validating pvrtex being able to fill this role. Then the example could still be put in to show how to use it.

**Pull request closed without merging**